backend/services/collab/src/index.ts

```diff
@@ -0,0 +1,158 @@
+  max: 1000,
+});
+app.use(limiter);
+
+/**
+ * Initialize CORS settings
+ */
+
+const cors = require("cors");
+const corsOptions = {
+  origin: process.env.ALLOW_ORIGIN_HEADER, 
+  credentials: true,
+  optionSuccessStatus: 200,
+  exposedHeaders: ["Set-Cookie"],
+};
+
+app.use(cors(corsOptions));
+
+/**
+ * Initialize cookie settings
+ * App will use cookie to obtain player info if player did not login
+ */
+
+
+const cookieParser = require('cookie-parser');
+app.use(cookieParser());
+
+/**
+ * Initialize MongoDB settings
+ */
+
+const mongoose = require('mongoose');
+mongoose.connect(process.env.DATABASE_URL, { useNewUrlParser: true });
+const db = mongoose.connection;
+db.on('error', (error: Error) => console.error(error));
+db.once('open', () => console.log('Connected to MongoDB'));
+
+/**
+ * Set up the url routes
+ */
+
+const pairingRouter = require('./routes/pairing');
+app.use('/pairing', pairingRouter);
+
+
+/**
+ * Start server on port
+ */
+app.listen(process.env.PORT, () => console.log('Server has started on port:', process.env.PORT));
+
+/**
+ * Setting up WebSocket
+ */
+
+const { WebSocketServer } = require('ws');
+const http = require('http');
+
+// Spinning the http server and the WebSocket server.
+const server = http.createServer();
+const wsServer = new WebSocketServer({ server });
+const port = process.env.WEBSOCKET_PORT;
+server.listen(port, () => {
+  console.log(`WebSocket server is running on port ${port}`);
+});
+
+// List of user: partner pairs
+const partners: { [userId: string]: string } = {};
+// List of clients' WebSocket connection
+const clients: { [userId: string]: WebSocket } = {};
+// Keeps track of whose turn it is in each pair
+const pairs: { [pairId: string]: string } = {};
+
+// A new client connection request received
+wsServer.on('connection', function(connection: WebSocket, request: Request, client) {
+  console.log(`Recieved a new connection.`);
+
+  // Store the new connection and handle messages
+  //console.log(`${connection} connected.`);
+  //console.log("Client ", client)
+
+  const params = getQueryParams(request.url);
+  //console.log("Query params ", params);
+  const pairId = params["pairId"];
+  const userId = params["userId"];
+
+  //console.log("Pair: ", pairId, "User: ", userId);
+
+  // Check pair exists, else close connection
+  getPair(pairId, userId).then(pairDoc => {
+    if (pairDoc !== null) {
+      if (pairDoc.isUser1Turn) {
+        pairs[pairId] = pairDoc.user1;
+      } else {
+        pairs[pairId] = pairDoc.user2;
+      }
+
+      const partnerId = userId === pairDoc.user1 ? pairDoc.user2 : pairDoc.user1;
+
+      partners[userId] = partnerId;
+      clients[userId] = connection;
+
+      // Check if partner has already connected
+      // If so, send READY to the pair
+      if (partnerId in clients) {
+        handleReady(connection, clients[partnerId]);
+      } 
+    } else {
+      console.log("Closing connection for user ", userId);
+      connection.close();
```

Comment:
do we want to delete from the pair from mongo for connections that are closed? otherwise the mongo will grow indefinitely (altho i dont think scalability is a factor in this project), just a suggestion

Reply:
Yes cos I was just testing stuff and didn't want to keep recreating the pair. But will delete that once everything working